### PR TITLE
MINOR: add Kafka Streams upgrade notes for 2.3 release

### PR DESCRIPTION
**Pull request by mjsax**
*More detailed description of your change,
if necessary. The PR title and PR message become
the squashed commit message, so use a separate
comment to ping reviewers.*

*Summary of testing strategy (including rationale)
for the feature or bug fix. Unit and/or integration
tests are expected for any behaviour change and
system tests should be considered for larger changes.*

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by mjsax** [comment]
Call for review @guozhangwang @bbejeck @vvcephei @ableegoldman @cadonna @abbccdda 

Tried to cover all KIP for this release. Hope I did not forget anything.

**Comment by ableegoldman** [comment]
Should we mention upgrading Rocks somewhere? Not an API change exactly but it exposes some new configs

File: docs/streams/upgrade-guide.html [hunk 1]
```diff
@@ -70,6 +70,48 @@ <h1>Upgrade Guide and API Changes</h1>
...
+        For more details please read <a href="https://cwiki.apache.org/confluence/display/KAFKA/KIP-443%3A+Return+to+default+segment.ms+and+segment.index.bytes+in+Streams+repartition+topics">KIP-443</a>.
+    </p>
+
+    <p>
+        To avoid memory leaks, <code>RocksDBConfigSetter</code> has a new <code>close()</code> method that is called on shutdown.
+        Users should implement this method to release any memory used by RocksDB config objects, by closing those objects.
+        For more details please read <a href="https://cwiki.apache.org/confluence/display/KAFKA/KIP-453%3A+Add+close%28%29+method+to+RocksDBConfigSetter">KIP-453</a>.
+    </p>
+
     <h3><a id="streams_api_changes_230" href="#streams_api_changes_230">Streams API changes in 2.3.0</a></h3>
```
Comment (ableegoldman) [inline docs/streams/upgrade-guide.html]: This was moved to above and should be deleted

File: docs/streams/upgrade-guide.html [hunk 2]
```diff
@@ -70,6 +70,48 @@ <h1>Upgrade Guide and API Changes</h1>
...
+        For more details please read <a href="https://cwiki.apache.org/confluence/display/KAFKA/KIP-258%3A+Allow+to+Store+Record+Timestamps+in+RocksDB">KIP-258</a>.
+    </p>
+
+    <p>
+        To improve type safety, a new operator <code>KStream#flatTransformValues</code> is added.
+        For more details please read <a href="https://cwiki.apache.org/confluence/display/KAFKA/KIP-313%3A+Add+KStream.flatTransform+and+KStream.flatTransformValues">KIP-313</a>.
+    </p>
+
+    <p>
+        In previous release, Kafka Streams only shipped with a built-in in-memory key-value store.
```
Comment (ableegoldman) [inline docs/streams/upgrade-guide.html]: These are mentioned again below (line 76), you should remove the other comment

File: docs/streams/upgrade-guide.html (hunk 1 above)
Comment (mjsax) [reply docs/streams/upgrade-guide.html]: Ups. Overlooked this.

**Comment by mjsax** [comment]
Updated this.

**Review by cadonna — APPROVED** [review]
LGTM

File: docs/streams/upgrade-guide.html [hunk 3]
```diff
@@ -71,14 +71,60 @@ <h1>Upgrade Guide and API Changes</h1>
...
+    <p>
+        In 2.3.0 we have added default implementation to <code>close()</code> and <code>configure()</code> for <code>Serializer</code>,
+        <code>Deserializer</code> and <code>Serde</code> so that they can be implemented by lambda expression.
+        For more details please read <a href="https://cwiki.apache.org/confluence/display/KAFKA/KIP-331+Add+default+implementation+to+close%28%29+and+configure%28%29+for+Serializer%2C+Deserializer+and+Serde">KIP-331</a>.
+    </p>
+
+    <p>
+        To improve operator semantics, new store types are added that allow storing an additional timestamp per key-value pair or window.
+        Some DSL operators (for example KTables) are using those new stores.
+        Hence, you can now receive the last update timestamp via Interactive Queries if you change the <code>QueryableStoreType</code>.
```
Comment (vvcephei) [inline docs/streams/upgrade-guide.html]:
```suggestion
        Hence, you can now retrieve the last update timestamp via Interactive Queries if you specify <code>TimestampedWindowStoreType</code> or <code>TimestampedKeyValueStoreType</code>.
```

File: docs/streams/upgrade-guide.html [hunk 4]
```diff
@@ -71,14 +71,60 @@ <h1>Upgrade Guide and API Changes</h1>
...
+        While this change is mainly transparent to the user, there are some corner cases that may require code changes:
```
Comment (vvcephei) [inline docs/streams/upgrade-guide.html]: Maybe a nit, but I feel it's a bit strange to call someone "the user" when you're talking directly to them.

```suggestion
        While this change is mainly transparent, there are some corner cases that may require code changes:
```

File: docs/streams/upgrade-guide.html [hunk 5]
```diff
@@ -71,14 +71,60 @@ <h1>Upgrade Guide and API Changes</h1>
...
+        <strong>Caution: If you receive an untyped store and use a cast, you might need to update your code to cast to the correct type.</strong>
```
Comment (vvcephei) [inline docs/streams/upgrade-guide.html]: Maybe we can drop in the actual exception, to improve the time-to-discover the reason, if someone is just searching:
`java.lang.ClassCastException: class org.apache.kafka.streams.state.ValueAndTimestamp cannot be cast to class ${actual value type}` upon _getting a value from the store_.

**Review by vvcephei — APPROVED** [review]
A couple of small suggestions. Thanks!

**Comment by mjsax** [comment]
Updated this.

File: docs/streams/upgrade-guide.html [hunk 6]
```diff
@@ -71,14 +71,64 @@ <h1>Upgrade Guide and API Changes</h1>
...
+    </p>
+
+    <p>
+        To improve type safety, a new operator <code>KStream#flatTransformValues</code> is added.
+        For more details please read <a href="https://cwiki.apache.org/confluence/display/KAFKA/KIP-313%3A+Add+KStream.flatTransform+and+KStream.flatTransformValues">KIP-313</a>.
+    </p>
+
+    <p>
+        Kafka Streams used to set the configuration parameter <code>max.poll.interval.ms</code> to <code>Interger.MAX_VALUE</code>.
+        In default value is removed and Kafka Streams uses the consumer default value now.
```
Comment (bbejeck) [inline docs/streams/upgrade-guide.html]: nit: 'In` -> `Its`

**Review by bbejeck — APPROVED** [review]
Just one very minor nit, otherwise LGTM

**Comment by mjsax** [comment]
Merged to `trunk` and cherry-picked to `2.3`

**Pull request merged**